stan/math/prim/fun/variance.hpp

```diff
@@ -37,28 +39,25 @@ inline return_type_t<T> variance(const std::vector<T>& v) {
  * Returns the sample variance (divide by length - 1) of the
  * coefficients in the specified matrix
  *
- * @tparam T type of elements in the vector
- * @tparam R number of rows in the matrix, can be Eigen::Dynamic
- * @tparam C number of columns in the matrix, can be Eigen::Dynamic
+ * @tparam EigMat type inheriting from `EigenBase` that does not have an `var`
+ *  scalar type.
  *
  * @param m matrix
  * @return sample variance of coefficients
  * @throw <code>std::invalid_argument</code> if the matrix has size zero
  */
-template <typename T, int R, int C>
-inline return_type_t<T> variance(const Eigen::Matrix<T, R, C>& m) {
+template <typename EigMat, require_eigen_t<EigMat>* = nullptr,
+          require_not_vt_var<EigMat>* = nullptr>
+inline value_type_t<EigMat> variance(EigMat&& m) {
+  using eig_value = value_type_t<EigMat>;
+  using ref_inner = const typename std::decay_t<EigMat>::PlainObject;
```

Comment:
```suggestion
  using T_plain = const plain_type_t<EigMat>;
```
I prefere naming variables by what they are, not how they will be used.